### ISSUE TITLE
[rollup-plugin] - Sourcemaps failing to upload when directory containing them has been deleted prior to upload happening

### ISSUE THREAD

**Issue by joekrump** [honeybadger-io/honeybadger-js#1053]
## What are the steps to reproduce this issue?
1. Set up an app which runs `vite build` in a [Heroku Nodejs buildpack](https://github.com/heroku/heroku-buildpack-nodejs)
2. Have the build command run in the buildpack

## What happens?
- Most sourcemap files that are output are uploaded
- Sourcemap files that are found in the ephemeral `tmp/build_...` directory are not uploaded because they no longer exist, even though their entries show up in the bundle object which this plugin reads entries from. The build fails because this results in an Error being thrown.

## What were you expecting to happen?
- No issues. Sourcemaps are uploaded successfully.

## Any logs, error output, etc?
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/3317231/228605068-bcc104e5-7e46-46dd-b560-eeae571fa9bd.png">

## What versions are you using?
**Operating System:** 
Building on the Heroku-20 stack
-----> Using buildpacks:
       1. heroku/nodejs
       2. heroku/ruby
**Package Name:** `@honeybadger-io/rollup-plugin`
**Package Version:** 5.1.5
**Browser Version:** N/A

**Comment by joekrump** [comment]
It would be great if I could set something like an `ignorePath` regex or glob which I could set to `"tmp/build_*"` in the plugin config options. That would then be able to be used in `isSourcemap`: https://github.com/honeybadger-io/honeybadger-js/blob/a10f68655e4949e32f0fbe0ed8cdb763232d6497/packages/rollup-plugin/src/rollupUtils.ts#LL21C25-L21C25

**Comment by BethanyBerkowitz** [comment]
Hi @joekrump, thanks for reaching out! Just to clarify, you believe the files in `tmp/` should not be uploaded, correct? I want to understand why they're even being generated since I don't see that happening when building locally with Vite. When I build, it looks like I just get assets in `/dist/assets`. Do you know if this is a Heroku thing?

Your suggestion of setting an `ignorePath` seems very reasonable, I'm just wondering if dumping files into `/tmp` and then deleting them is expected behavior.. if so, perhaps we should exclude those files by default rather than adding additional config. 

Are you able to tell what folder those assets that were successfully uploaded were located in? 

Also, would you mind sharing your `vite.config.js` just in case there's anything relevant?

**Comment by subzero10** [comment]
Hey @joekrump, thanks for the update and the config file.

I have a question:
- Those files that are referenced in the `tmp` folder, aren't they necessary files to be uploaded to Honeybadger? I mean, if we went ahead with an `ignorePath` option or ignored files in `tmp` folder by default, you would be missing source maps, correct?

Additionally, a comment on your `vite.config.ts` file:
- You define `HONEYBADGER_API_KEY` and `HONEYBADGER_REVISION`, but I think they are irrelevant since you set `apiKey` and `revision` in `honeybadgerPluginOptions`. Is this intended?

Finally, I think I have all I need to try this on my side and see if I can reproduce your issue and debug further.

**Comment by joekrump** [comment]
Re `vite.config.ts`. I've defined `HONEYBADGER_API_KEY`, etc because they are used elsewhere in some other js code where I'm calling `Honeybadger.configure()` when my vue app first mounts.

**Comment by joekrump** [comment]
The files in `tmp` get copied over into my `public` directory and so all the files will end up existing there.

**Comment by subzero10** [comment]
Hey @joekrump, I created a very simple vite app and deployed on Heroku. I couldn't reproduce the errors shown above.

In any case, I will go ahead and create an `ignorePath` configuration option, so that you can continue using the plugin without these errors.

**Comment by joekrump** [comment]
Thanks @subzero10?! Any chance you can make that sample app open source that sample vite app? It might help me out and others in the future.

**Comment by subzero10** [comment]
> Thanks @subzero10! Any chance you can make that sample app open source that sample vite app? It might help me out and others in the future.

[Here](https://github.com/subzero10/vite-js-1053) it is.
Note that it doesn't even create a static page. I just added the nodejs buildpack on Heroku and made sure that the build step is executed:
```
-----> Build
       Running build
       
       > vite-js-1053@0.0.1 build
       > vue-tsc && vite build
       
       vite v4.3.1 building for production...
       transforming...
       ✓ 18 modules transformed.
       rendering chunks...
       computing gzip size...
       dist/index.html                  0.45 kB │ gzip:  0.30 kB
       dist/assets/vue-5532db34.svg     0.50 kB │ gzip:  0.31 kB
       dist/assets/index-c322ae43.css   1.30 kB │ gzip:  0.67 kB
       dist/assets/index-2424b2d8.js   54.54 kB │ gzip: 22.02 kB │ map: 565.12 kB
       Successfully uploaded assets/index-2424b2d8.js.map to Honeybadger
       1 sourcemap file(s) successfully uploaded to Honeybadger
       ✓ built in 1.54s
       
-----> Caching build
       - npm cache
       
-----> Pruning devDependencies
       
       up to date, audited 22 packages in 287ms
       
       2 packages are looking for funding
         run `npm fund` for details
       
       found 0 vulnerabilities
       npm notice 
       npm notice New minor version of npm available! 9.5.1 -> 9.6.5
       npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.6.5>
       npm notice Run `npm install -g npm@9.6.5` to update!
       npm notice 
       
-----> Build succeeded!
 !     This app may not specify any way to start a node process
       https://devcenter.heroku.com/articles/nodejs-support#default-web-process-type
-----> Discovering process types
       Procfile declares types     -> (none)
       Default types for buildpack -> web
-----> Compressing...
       Done: 45.7M
-----> Launching...
       Released v4
       https://vite-js-1053.herokuapp.com/ deployed to Heroku
```

**Comment by joekrump** [comment]
@subzero10 something I just caught regarding the upload errors.

Error output:

```
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/errors.f2c021a1.js'
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/quill.32217bd5.js'
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/styleguide.11183cda.js'
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/users_search.b38a9299.js'
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/application.31b9c4cb.js'
 Error: ENOENT: no such file or directory, open '/tmp/build_40af8e8b/public/vite/assets/devise.8cbd75cd.js'
```

Those failed files all seem to correspond to the `.sass` files I have:

<img width="217" alt="image" src="https://user-images.githubusercontent.com/3317231/234101415-3c36e18d-8087-4a30-9dd9-bd044e94d879.png">

I can see that the `.js` files in the directory had sourcemaps uploaded successfully:

```
Successfully uploaded assets/quill.b96c7531.js.map to Honeybadger
Successfully uploaded assets/application.1d96dddf.js.map to Honeybadger
```

**Comment by joekrump** [comment]
So I'm starting to think that the root cause of the issue might be something to do with what I've outlined in this other issue: https://github.com/ElMassimo/vite_ruby/issues/358

**Comment by joekrump** [comment]
If that's the case, the ignorePath could still be helpful, but what might be even more helpful might be something like an `ignoreSourcemapFn` or something like that which would allow someone to run some custom check on a sourcemap before determining it's something that should be uploaded or not. What do you think?

**Comment by subzero10** [comment]
Hey @joekrump, this is great info! I am working already on adding support for `ignorePath` and I will also work on your latest findings. However, I am thinking that if `sourcesContent` is empty, we should probably ignore those files by default.

**Comment by joekrump** [comment]
>I am thinking that if `sourcesContent` is empty, we should probably ignore those files by default.

That would be a great default to add and would fix my issue I think.

**Comment by subzero10** [comment]
Hey @joekrump, we just released `v5.4.0` of the `@honeybadger-io/rollup-plugin` with an `ingorePaths` option and an algorithm improvement to skip source maps that have empty `sourcesContent`.

Thank you for all your help on this!

cc @BethanyBerkowitz

**Comment by joekrump** [comment]
🙌🏻 Amazing! Thanks for you work on this @subzero10 and @BethanyBerkowitz
